examples/gno.land/p/demo/grc/grc20/token.gno

```diff
@@ -2,44 +2,234 @@ package grc20
+// SpendAllowance decreases the allowance of the specified owner and spender.
+func (led *AdminLedger) SpendAllowance(owner, spender std.Address, amount uint64) error {
+	if !owner.IsValid() {
+		return ErrInvalidAddress
+	}
+	if !spender.IsValid() {
+		return ErrInvalidAddress
+	}
+
+	currentAllowance := led.allowance(owner, spender)
+	if currentAllowance < amount {
+		return ErrInsufficientAllowance
+	}
+
+	key := allowanceKey(owner, spender)
+	newAllowance := currentAllowance - amount
+
+	if newAllowance == 0 {
+		led.allowances.Remove(key)
+	} else {
+		led.allowances.Set(key, newAllowance)
+	}
+
+	return nil
 }
 
-func (t *token) TransferFrom(from, to std.Address, amount uint64) error {
-	spender := std.PrevRealm().Addr()
-	if err := t.banker.SpendAllowance(from, spender, amount); err != nil {
+// Transfer transfers tokens from the specified from address to the specified to address.
+func (led *AdminLedger) Transfer(from, to std.Address, amount uint64) error {
+	if !from.IsValid() {
+		return ErrInvalidAddress
+	}
+	if !to.IsValid() {
+		return ErrInvalidAddress
+	}
+	if from == to {
+		return ErrCannotTransferToSelf
+	}
+
+	var (
+		toBalance   = led.balanceOf(to)
+		fromBalance = led.balanceOf(from)
+	)
+
+	if fromBalance < amount {
+		return ErrInsufficientBalance
+	}
+
+	var (
+		newToBalance   = toBalance + amount
+		newFromBalance = fromBalance - amount
+	)
+
+	led.balances.Set(string(to), newToBalance)
+	led.balances.Set(string(from), newFromBalance)
+
+	std.Emit(
+		TransferEvent,
+		"from", from.String(),
+		"to", to.String(),
+		"value", strconv.Itoa(int(amount)),
+	)
+
+	return nil
+}
+
+// TransferFrom transfers tokens from the specified owner to the specified to address.
+// It first checks if the owner has sufficient balance and then decreases the allowance.
+func (led *AdminLedger) TransferFrom(owner, spender, to std.Address, amount uint64) error {
+	if led.balanceOf(owner) < amount {
+		return ErrInsufficientBalance
+	}
+	if err := led.SpendAllowance(owner, spender, amount); err != nil {
 		return err
 	}
-	return t.banker.Transfer(from, to, amount)
+	// XXX: since we don't "panic", we should take care of rollbacking spendAllowance if transfer fails.
+	return led.Transfer(owner, to, amount)
+}
+
+// Approve sets the allowance of the specified owner and spender.
+func (led *AdminLedger) Approve(owner, spender std.Address, amount uint64) error {
+	if !owner.IsValid() || !spender.IsValid() {
+		return ErrInvalidAddress
+	}
+
+	led.allowances.Set(allowanceKey(owner, spender), amount)
+
+	std.Emit(
+		ApprovalEvent,
+		"owner", string(owner),
+		"spender", string(spender),
+		"value", strconv.Itoa(int(amount)),
+	)
+
+	return nil
+}
+
+// Mint increases the total supply of the token and adds the specified amount to the specified address.
+func (led *AdminLedger) Mint(address std.Address, amount uint64) error {
+	if !address.IsValid() {
+		return ErrInvalidAddress
+	}
+
+	// XXX check for overflow
```

Comment:
can use math/overflow here and in general

Reply:
37e70d48d